---
title: "Cell expression"
author: "Panteleev Dmitrii"
output: html_document
---


```{r libraries, include=FALSE}
library(readxl)
library(limma)
library(ape)
library(dendextend)
library(RColorBrewer)
library(gprofiler2)
library(pvclust)
library(gplots)
library(NMF)
library(vegan)
library(Biobase)
library(DT)
library(ggplot2)
library(graphlayouts)

library(impute)
library(ggrepel)
library(enrichplot)
```

# Download data

We don't need to eliminate the whole samples(columns) because none of them have all empty observations(or significant part of them contains N/A). 

```{r include=FALSE}
initial <- read.csv('peaks_data.csv', sep=',')
# rownames(initial) <- initial$Gene_id # we cannot do it because we have some values with the same name

cells <- initial[,-c(1,2)]


cells[cells$Gene_id == 'GNAS',] # both of them have identical result 
cells[cells$Gene_id == 'NACA',] # the complex case, but i think i can delete the second row, because it have more NA values than the first one.
cells[cells$Gene_id == 'nan',] # I think that i can eleminate these strings because names of genes are unknown
cells[cells$Gene_id == 'PALM2AKAP2',] # the second row has more than half of the NA values
cells <- cells[-c(848,1322, 102), ]
cells <- cells[cells$Gene_id != '' & cells$Gene_id != 'nan',]



rownames(cells) <- cells$Gene_id

cells <- cells[,-1]
```

## EDA

Remove genes with half and more missing values

```{r}
genes_with_NA <- names(which(rowSums(is.na(cells)) > ncol(cells)/2))
cells_without_NA <- cells[!rownames(cells) %in% genes_with_NA,]
```

The result: We filtered 1311 genes.

Imputation with a KNN method:

```{r}
cells_without_NA <- as.matrix(cells_without_NA)
cells_without_NA_trans <- t(cells_without_NA)
knn_cells <- impute.knn(cells_without_NA_trans, k = 5)
knn_cells <- knn_cells$data
knn_cells <- as.data.frame(knn_cells)

knn_cells <- t(as.matrix(knn_cells))


cells_experiment <- as.factor(c("1", "1", "1", "1","1", "1", "1", "1", "1", "1", "1", "1", "1", "1", "2", "2", "2", "2", "2", "2", "2", "2", "2", "2", "2", "2", "2" , "2", "2", "2", "2", "2", "2"))
```

```{r}
pal <- brewer.pal(n = 9, name = "Set1")
cols <- pal[cells_experiment]
boxplot(knn_cells, outline = FALSE, main = "Initial data", col = cols)
legend("topright", levels(cells_experiment), fill = pal, bty = "n", xpd = T)
```

Data normalization:

```{r}
cells_log <- log2(knn_cells)
cells_norm <- normalizeQuantiles(as.matrix(cells_log))
boxplot(cells_norm, outline = FALSE, main = "Normalised data", col = cols)
legend("topright", levels(cells_experiment), fill = pal, bty = "n", xpd = T)
```
```{r}
colSums(is.infinite(cells_norm))
```

We can see that Median and Mean of box plots have became more similar than before a normalization step, but also we can notice that we have a slight fluctuation between samples at this plot. 
This may be due to the null values of knn_cells and -inf-values of cells_nnm as a result of logarithm.
That is why I decided to add +1 to eliminate -inf:

```{r}
cells_log <- log2(knn_cells+1)
cells_norm <- normalizeQuantiles(as.matrix(cells_log))
boxplot(cells_norm, outline = FALSE, main = "Normalised data", col = cols)
legend("topright", levels(cells_experiment), fill = pal, bty = "n", xpd = T)
```
This result is more better then previous one, because our cells_norm hasn't -inf data.
```{r}
colSums(is.infinite(cells_norm))
```

```{r}
cells_pca <- t(cells_norm)
terapod_pca <- rda(cells_pca, scale = TRUE)

rownames(cells_pca) <- c("BT", "BT","BT", "BT","BT", "BT","BT", "BT","BT", "BT","BT", "BT","BT", "BT", "CJK", "CJK", "CJK", "CJK", "CJK", "CJK","CJK", "CJK", "CJK", "CJK","CJK", "CJK", "CJK", "CJK", "CJK", "CJK","CJK", "CJK", "CJK")


df_scores <- data.frame(cells_pca,
                        scores(terapod_pca, display = "sites", choices = c(1, 2, 3), scaling = "sites"))

p_scores <- ggplot(df_scores, aes(x = PC1, y = PC2)) + 
  geom_point(aes(color = rownames(cells_pca)), alpha = 0.5) +
  coord_equal(xlim = c(-3, 3), ylim = c(-3, 3)) + ggtitle(label = "Ordination") + theme_bw()
p_scores
```



MA-plot for a comparison of groups

```{r}
maplot <- function(X1, X2, pch = 21, main = "MA-plot", xlab = "Average log-expression", ylab = "Expression log-ratio", lpars = list(col = "blue", lwd = 2), ...){
  
  X <- (rowMeans(X2) + rowMeans(X1)) / 2
  Y <- rowMeans(X2) - rowMeans(X1)
  
  scatter.smooth(x = X, y = Y, main = main, pch = pch, xlab = xlab, ylab = ylab, lpars = lpars, ...)
  abline(h = c(-1, 0, 1), lty = c(2, 1, 2))
}
```


MA plot: CJK versus
```{r}
maplot(cells_log[,1:14], cells_log[,15:33],  main = "Non-normalized data")
```




```{r}
maplot(cells_norm[,1:14], cells_norm[,15:33],  main = "Normalized data")
```


```{r}
maplot(cells_norm[,1:14], cells_norm[,c(15:20,24:33)],  main = "Normalized data")
```


```{r}
#cells_norm[,21:23]
```


I don't see the significant difference between plot, Normalised graph is less curved. The normalization step didn't impact on it strongly. We can excude problematic columns and a line can be less curved.   

____________________________________________________________________________________





### Dif expression

```{r}

expr_data <- as.matrix(cells_norm)


pheno_data <- data.frame(cells_experiment)
rownames(pheno_data) <- colnames(cells_norm)
pheno_metadata <- data.frame(
  labelDescription = c("Experimental condition"),
  row.names=c("Condition"))
pheno_data <- new("AnnotatedDataFrame",
                 data = pheno_data,
                 varMetadata = pheno_metadata)

feature_data <- data.frame(Prot = rownames(expr_data))
rownames(feature_data) <- rownames(expr_data)
feature_metadata <- data.frame(
  labelDescription = c("Protein name"),
  row.names = c("Protein"))
f_data <- new("AnnotatedDataFrame",
              data = feature_data,
              varMetadata = feature_metadata)


exp_set <-
  ExpressionSet(assayData = expr_data,
                phenoData = pheno_data,
                featureData = f_data)
```

We got ExpressionSet with all our data.


```{r}
X <- model.matrix(~ cells_experiment, pData(exp_set))
fit <- lmFit(exp_set, design = X, method = "robust", maxit = 1000)

efit <- eBayes(fit)
```

We performed  Empirical Bayes statistics and now we have differentially expressed proteins

```{r}
MA_limma <- function(efit, coef, n = 10, signif = TRUE, fdr = 0.05, lfc = 0, text = TRUE, cex.text = 0.8, col.text = "grey20", main = "MA-plot", xlab = "Average log-expression", ylab = "Expression log-ratio", pch = 19, pch.signif = 21, col = "darkgreen", alpha = 0.3, cex = 0.3, ...){
  
  R <- efit$coefficients[, coef]
  I <- efit$Amean

  col_btransp <- adjustcolor(col, alpha.f = alpha)
  
  plot(I, R, cex = cex, main = main, pch = pch, xlab = xlab, ylab = ylab, col = col_btransp, ...)
  abline(h = 0)
  
  if(signif){
    sign <- p.adjust(efit$p.value[, coef], method = "BH") <= fdr
    large <- abs(efit$coefficients[, coef]) >= lfc
    points(I[sign & large], R[sign & large], cex = cex*2, col = "orange2", pch = pch.signif)
  }
  
  if(text){
    ord <- order(efit$lods[, coef], decreasing = TRUE)
    top_n <- ord[1:n]
    text(I[top_n], R[top_n], labels = efit$genes[top_n, ], pos = 4, cex = cex.text, col = col.text)
  }
}


MA_limma(efit, coef = 2, n = 30)
```

We can see at this MA plot dif exp proteins like orange points and the name of first 30 protein.

```{r}
# We can get top 100 proteins for the further study 

my_list <- topTable(efit, coef = 2, n = 100)

dif_exp_set <- exp_set[fData(exp_set)$Prot %in% my_list$Prot, ]
```


```{r fig.height=10, fig.width=15}
dat <- as.matrix(exprs(dif_exp_set))

pal_blue_red <- colorpanel(75, low = "steelblue", mid = "black", high = "red")
heatmap.2(dat, col = pal_blue_red, scale = "row", key = TRUE, symkey = FALSE, density.info = "none", trace = "none", cexRow = 0.9, cexCol = 1, margins = c(4, 3), keysize = 0.8, key.par = list(mar = c(3, 0.1, 3, 0.1)))
```


We can see different patterns between BT and CJK samples. It's a good result. We can notice an explicit group of that up or down regulated in BT samples.

```{r}
topTable(efit, coef = 2)
numGenes <- nrow(exprs(exp_set))
full_list <- topTable(efit, number = numGenes)
full_list <- full_list[full_list$adj.P.Val <= 0.05,]
write.csv(full_list, 'C:/Users/matt/Desktop/proteomicsHW/DE_full.csv')
```


### MA-plot of the first 20 differentially expressed proteins, but such that the expression level differs with  two times

```{r}

MA_limma(efit, coef = 2, n = 80, text = F, lfc = 1)
```



```{r}

my_list <- full_list

dif_exp_set <- exp_set[fData(exp_set)$Prot %in% my_list$Prot, ]
```

# Volcano plot
```{r }
volcano_list <- full_list

volcano1 <- ggplot(data = volcano_list, aes(x = logFC, y = -log10(adj.P.Val))) + geom_point()

volcano2 <- volcano1 + geom_vline(xintercept=c(-1, 1), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")

volcano_list$diffexpressed <- "NO"       
```


```{r }
volcano_list$diffexpressed[volcano_list$logFC > 1 & volcano_list$adj.P.Val < 0.05] <- "UP"
```


```{r }
volcano_list$diffexpressed[volcano_list$logFC < -1 & volcano_list$adj.P.Val< 0.05] <- "DOWN"

volcanodif1 <- ggplot(data = volcano_list, aes(x = logFC, y = -log10(adj.P.Val), col = diffexpressed)) + geom_point() + theme_minimal()
```

 Add lines as before...

```{r }
volcanodif2 <- volcanodif1 + geom_vline(xintercept=c(-1, 1), col="red") +
  geom_hline(yintercept=-log10(0.05), col="red")

volcano_list$Prot <- as.character(volcano_list$Prot)
volcano_list$delabel <- NA
volcano_list$delabel[volcano_list$diffexpressed != "NO"] <- volcano_list$Prot[volcano_list$diffexpressed != "NO"]
#volcano_list[volcano_list$adj.P.Val<=0.6,]$delabel <- volcano_list[volcano_list$adj.P.Val<=0.6,]$Prot

plot_proteins <- ggplot(data=volcano_list, aes(x=logFC, y=-log10(adj.P.Val), col=diffexpressed, label=delabel)) + 
  geom_point() + 
  theme_minimal() +
  geom_text_repel(size=3, colour = 'black', max.overlaps = 30)

plot_final <- plot_proteins + geom_vline(xintercept=c(-1, 1), col="red") + geom_hline(yintercept=-log10(0.05), col="red")
plot_final
```

Here we can see blue, red and greem points, according to up, dowm and non-regulated proteins.

# GO enrichment analysis
In order to group significantly differentially expressed proteins we have used GO enrichment analysis.

```{r warning=FALSE, message=FALSE}
# keep only the significant proteins results
sig <- subset(full_list, adj.P.Val < 0.05)
# get the significant up-regulated proteins
up <- subset(sig, logFC > 0)
# get the significant down-regulated proteins
down <- subset(sig, logFC < 0)
```

```{r warning=FALSE, message=FALSE}
# needed to convert to enrichResult object
up_names <- gconvert(row.names(up))
down_names <- gconvert(row.names(down))
```

## Up-regulated proteins
```{r warning=FALSE, message=FALSE}
# enrichment analysis using proteins names
multi_gp_up_reg <- gost(list("up-regulated" = up_names$name), multi_query = FALSE, evcodes =TRUE)
# modify the g:Profiler data frame
gp_mod_up = multi_gp_up_reg$result[, c("query", "source", "term_id","term_name", "p_value", "query_size", "intersection_size", "term_size", "effective_domain_size", "intersection")]
gp_mod_up <- gp_mod_up[order(gp_mod_up$p_value), ]
gp_mod_up_BP <- gp_mod_up[gp_mod_up$source == "GO:BP", ]
gp_mod_up_BP$GeneRatio <- paste0(gp_mod_up_BP$intersection_size,  "/", gp_mod_up_BP$query_size)
gp_mod_up_BP$BgRatio <- paste0(gp_mod_up_BP$term_size, "/", gp_mod_up_BP$effective_domain_size)
names(gp_mod_up_BP) <- c("Cluster", "Category", "ID", "Description", "p.adjust", "query_size", "Count", "term_size", "effective_domain_size", "geneID", "GeneRatio", "BgRatio")
gp_mod_up_BP$geneID <- gsub(",", "/", gp_mod_up_BP$geneID)
row.names(gp_mod_up_BP) <- gp_mod_up_BP$ID
gp_mod_enrich_up_BP <- new("enrichResult", result = gp_mod_up_BP)
```

Draw enrichment plot:
```{r warning=FALSE, message=FALSE}
enrichplot::dotplot(gp_mod_enrich_up_BP, showCategory = 10) + ggplot2::labs(title = "up-regulated") + ggplot2::scale_color_gradient(low = "lightseagreen", high = "darkorange1")
```

## Down-regulated proteins
```{r warning=FALSE, message=FALSE}
# enrichment analysis using gene names
multi_gp_down_reg <- gost(list("down-regulated" = down_names$name), multi_query = FALSE, evcodes =TRUE)
# modify the g:Profiler data frame
gp_mod_down = multi_gp_down_reg$result[, c("query", "source", "term_id","term_name", "p_value", "query_size", "intersection_size", "term_size", "effective_domain_size", "intersection")]
gp_mod_down <- gp_mod_down[order(gp_mod_down$p_value), ]
# BP
gp_mod_down_BP <- gp_mod_down[gp_mod_down$source == "GO:BP", ]
gp_mod_down_BP$GeneRatio <- paste0(gp_mod_down_BP$intersection_size,  "/", gp_mod_down_BP$query_size)
gp_mod_down_BP$BgRatio <-  paste0(gp_mod_down_BP$term_size, "/", gp_mod_down_BP$effective_domain_size)
names(gp_mod_down_BP) <- c("Cluster", "Category", "ID", "Description", "p.adjust", "query_size", "Count", "term_size", "effective_domain_size", "geneID", "GeneRatio", "BgRatio")
gp_mod_down_BP$geneID <- gsub(",", "/", gp_mod_down_BP$geneID)
gp_mod_enrich_down_BP <- new("enrichResult", result = gp_mod_down_BP)
```

Draw enrichment plot:
```{r warning=FALSE, message=FALSE}
enrichplot::dotplot(gp_mod_enrich_down_BP, showCategory = 10) + ggplot2::labs(title = "down-regulated") + ggplot2::scale_color_gradient(low = "lightseagreen", high = "darkorange1")
```


Here we can see 2 lists of the most up and down regulated proteins.  



Biological explanation: 
We can see the interesting result that some gene groups is up and down regulated simultaneously (For instance, organonitrogen compound metabolic process, intracellular transport). We can notice that these two lists includes very simmilar group(for example, localisation and cellular localisation).
It can be explained that The Calcific Aortic Valve Disease is characterised thickening and calcification of the aortic valve leaflets. For this process cells must be rebuilt and that is why we see that some of transport, regulation and even organisation genes are down regulated but another ones are up regulated. 

